### PR TITLE
Make ExecTainted easier to extend

### PR DESCRIPTION
**Pull request by owen-mc**
No change in behaviour - this is just a refactoring. To add a method that executes a command, you can now define a class
extending ExecMethod.

### CODE REVIEW

**Review by smowton — COMMENTED** [review]
You should generalise this a little: let `ExecCallable` extend `Callable` instead of `Method` (then `ProcessBuilder`'s constructor can be an instance), and give it an abstract `getAnExecutedArgument`, which all overrides will currently define as `getArgument(0)`, though e.g. `myExec(int flags, String command)` would define as `getArgument(1)`.

File: java/ql/src/semmle/code/java/JDK.qll [hunk 1]
```diff
@@ -180,21 +180,36 @@ class TypeFile extends Class {
 /**
  * Any of the methods named `command` on class `java.lang.ProcessBuilder`.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/JDK.qll:181]: qldoc needs fixing

File: java/ql/src/semmle/code/java/frameworks/apache/Exec.qll [hunk 2]
```diff
@@ -6,16 +6,20 @@ library class TypeCommandLine extends Class {
   TypeCommandLine() { hasQualifiedName("org.apache.commons.exec", "CommandLine") }
 }
 
-library class MethodCommandLineParse extends ExecMethod {
+library class MethodCommandLineParse extends Method, ExecCallable {
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/apache/Exec.qll:9]: Let's get rid of the archaic `library` annotations while we're editing this file. And a few qldoc additions seems like easy inclusions as well.

File: java/ql/src/semmle/code/java/security/ExternalProcess.qll [hunk 3]
```diff
@@ -6,7 +6,9 @@ import semmle.code.java.frameworks.apache.Exec
 /**
  * A method that executes a command.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/ExternalProcess.qll:7]: qldoc needs fixing

File: java/ql/src/semmle/code/java/frameworks/apache/Exec.qll (hunk 2 above)
Comment (owen-mc) [reply java/ql/src/semmle/code/java/frameworks/apache/Exec.qll:9]: Shall I replace `library` with `private`?

File: java/ql/src/semmle/code/java/security/ExternalProcess.qll [hunk 4]
```diff
@@ -15,15 +17,10 @@ abstract class ExecMethod extends Method { }
...
-    )
-    or
-    exists(ConstructorCall expr, Constructor cons |
-      expr.getConstructor() = cons and
-      cons.getDeclaringType().hasQualifiedName("java.lang", "ProcessBuilder") and
-      expr.getArgument(0) = this
+    exists(Call execCall, ExecCallable execCallable, int i |
+      execCall.getArgument(i) = this and
+      execCallable = execCall.getCallee() and
+      i = execCallable.getAnExecutedArgument()
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/ExternalProcess.qll:23]: Please check locally that the join-order of this looks reasonable, i.e. execute a query that depends on this predicate and check in the log that it doesn't start the pipeline with a join on `i`.

File: java/ql/src/semmle/code/java/frameworks/apache/Exec.qll (hunk 2 above)
Comment (aschackmull) [reply java/ql/src/semmle/code/java/frameworks/apache/Exec.qll:9]: If the class isn't referenced anywhere else (i.e. things still compile), then yes, I think that's fine (even though it technically is a breaking change).

File: java/ql/src/semmle/code/java/security/ExternalProcess.qll (hunk 4 above)
Comment (owen-mc) [reply java/ql/src/semmle/code/java/security/ExternalProcess.qll:23]: You're right, it joins on `i` first. Adding `pragma[only_bind_into]` in the expected place fixed it.

File: java/ql/src/semmle/code/java/security/ExternalProcess.qll [hunk 5]
```diff
@@ -3,28 +3,27 @@ import semmle.code.java.Member
 import semmle.code.java.JDK
 import semmle.code.java.frameworks.apache.Exec
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/ExternalProcess.qll:4]: I don't think anything from these imports is directly referenced anymore, which means that they're just here to activate the abstract class instances. If that's the case then we should make them private and nest them in a private module to make it explicit that they're only imported for their class activation side effects.
```suggestion

private module Instances {
  private import semmle.code.java.JDK
  private import semmle.code.java.frameworks.apache.Exec
}
```

**Review by aschackmull — COMMENTED** [review]
One more inline comment, otherwise LGTM. Needs autoformat of a couple of files.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**